### PR TITLE
Add SPSA `gradient_transform`

### PR DESCRIPTION
**Pull request by dwierichs**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
PennyLane has the `SPSAOptimizer`, but no separate method to compute the gradient alone based on the SPSA idea.

**Description of the Change:**
This PR adds the differentiation method SPSA, i.e a single estimation of a gradient via simultaneous perturbation of all parameters and stochastic approximation.

**Benefits:**
New differentiation method without having to use `SPSAOptimizer`.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: pennylane/gradients/gradient_transform.py [hunk 1]
```diff
@@ -46,11 +46,11 @@ def gradient_analysis(tape, use_graph=True, grad_fn=None):
             the cached gradient analysis will be used.
     """
     # pylint:disable=protected-access
-    if grad_fn is not None and getattr(tape, "_gradient_fn", None) is grad_fn:
-        # gradient analysis has already been performed on this tape
-        return
-
     if grad_fn is not None:
+        if getattr(tape, "_gradient_fn", None) is grad_fn:
```
Comment (rmoyard) [inline pennylane/gradients/gradient_transform.py:50]: Nice move!

File: pennylane/gradients/spsa_gradient.py [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+    `Rademacher distribution. <https://en.wikipedia.org/wiki/Rademacher_distribution>`_
+    """
+    # pylint: disable=unused-argument
+    direction = np.zeros(num_params)
+    direction[indices] = np.random.choice([-1, 1], size=len(indices))
+    return direction
+
+
+@gradient_transform
+def _spsa_grad_new(
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:53]: Thank you so much for thinking about that!

File: pennylane/gradients/spsa_gradient.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+            gradient_tapes.append(tape)
+            extract_r0 = True
+
+        # Skip the unshifted tape
+        shifts = shifts[1:]
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    indices = [
+        i for i, _ in enumerate(tape.trainable_params) if (i in method_map and method_map[i] != "0")
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:260]:
```suggestion
        i for i in range(len(tape.trainable_params)) if (i in method_map and method_map[i] != "0")
```
That should work as well

File: pennylane/gradients/spsa_gradient.py [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+
+    tapes_per_grad = len(shifts)
+    all_coeffs = []
+    for idx_rep in range(num_directions):
+        direction = sampler(indices, len(tape.trainable_params), idx_rep)
+        inv_direction = qml.math.divide(
+            1, direction, where=(direction != 0), out=qml.math.zeros_like(direction)
+        )
+        # Use only the non-zero part of `direction` for the shifts, to skip redundant zero shifts
+        _shifts = qml.math.tensordot(h * shifts, direction[indices], axes=0)
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:271]: So here we select the directions that have trainable parameters?

File: pennylane/gradients/spsa_gradient.py [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+        all_coeffs.append(qml.math.tensordot(coeffs / h**n, inv_direction, axes=0))
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:272]: Why do we use the inverse for the main coeff and not directions?

File: pennylane/gradients/spsa_gradient.py [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+        shots or a single component of a shot vector"""
+
+        r0, results = (results[0], results[1:]) if extract_r0 else (f0, results)
+        if len(tape.measurements) == 1:
+            grads = 0
+            for rep, _coeffs in enumerate(all_coeffs):
+                res = results[rep * tapes_per_grad : (rep + 1) * tapes_per_grad]
+                if r0 is not None:
+                    res.insert(0, r0)
+                res = qml.math.stack(res)
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:287]: I am worried with stacking here. Is it possible to have maybe different shaped arrays at this point? It would trigger a ragged array warning.

File: pennylane/gradients/spsa_gradient.py [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+            checking.
+        shots (None, int, list[int], list[ShotTuple]): The device shots that will be used to
+            execute the tapes outputted by this transform. Note that this argument doesn't
+            influence the shots used for tape execution, but provides information
+            to the transform about the device shots and helps in determining if a shot
+            sequence was used to define the device shots for the new return types output system.
+        num_directions (int): Number of sampled simultaneous perturbation vectors. An estimate for
+            the gradient is computed for each vector using the underlying finite-difference
+            method, and afterwards all estimates are averaged.
+
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:107]: We are missing some documentation for the sampler

File: pennylane/gradients/spsa_gradient.py [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+      tensor(-0.19280803, requires_grad=True),
+      tensor(0.19280803, requires_grad=True)),
+     (tensor(0.34786926, requires_grad=True),
+      tensor(0.34786926, requires_grad=True),
+      tensor(-0.34786926, requires_grad=True)))
+
+    Note that the SPSA gradient is a statistical estimator that uses a given number of
+    function evaluations that does not depend on the number of parameters. While this
+    bounds the cost of the estimation, it also implies that the returned values are not
+    exact (even for devices with ``shots=None``) and that they will fluctuate.
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:142]: Should we provide a way to fix the randomness?

File: pennylane/gradients/spsa_gradient.py [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+        >>> dev = qml.device("default.qubit", wires=2, shots=shots)
+        >>> @qml.qnode(dev)
+        ... def circuit(params):
+        ...     qml.RX(params[0], wires=0)
+        ...     qml.RY(params[1], wires=0)
+        ...     qml.RX(params[2], wires=0)
+        ...     return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0))
+        >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+        >>> qml.gradients.spsa_grad(circuit, shots=shots, h=1e-2)(params)
+        (((array(0.), array(0.), array(0.)), (array(0.), array(0.), array(0.))),
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:217]: Great 💯

File: pennylane/gradients/spsa_gradient.py [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+            checking.
+        shots (None, int, list[int], list[ShotTuple]): The device shots that will be used to
+            execute the tapes outputted by this transform. Note that this argument doesn't
+            influence the shots used for tape execution, but provides information
+            to the transform about the device shots and helps in determining if a shot
+            sequence was used to define the device shots for the new return types output system.
+        num_directions (int): Number of sampled simultaneous perturbation vectors. An estimate for
+            the gradient is computed for each vector using the underlying finite-difference
+            method, and afterwards all estimates are averaged.
+
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:390]: Also sampler missing.

File: tests/gradients/test_spsa_gradient.py [hunk 11]
```diff
@@ -0,0 +1,878 @@
...
+
+        for _ in range(5):
+            direction = _rademacher_sampler(ids, num)
+            assert direction.shape == (num,)
+            assert set(direction).issubset({0, -1, 1})
+            assert np.allclose(np.abs(direction)[ids_mask], 1)
+            assert np.allclose(direction[~ids_mask], 0)
+
+    @pytest.mark.parametrize("num_directions, tol", [(100, 0.1), (500, 5e-3)])
+    def test_convergence_single_par(self, num_directions, tol):
```
Comment (rmoyard) [inline tests/gradients/test_spsa_gradient.py:55]: Should it not be in the other class, I think we should keep it for Rademacher unit tests and not integration.

File: tests/returntypes/test_spsa_gradient_new.py [hunk 12]
```diff
@@ -0,0 +1,1052 @@
...
+        x = np.random.rand(3)
+        circuits = [qml.QNode(cost, dev) for cost in (cost1, cost2, cost3, cost4, cost5, cost6)]
+
+        transform = [qml.math.shape(spsa_grad(c)(x)) for c in circuits]
+
+        expected = [(3,), (3,), (2, 3), (3, 4), (3, 4), (2, 3, 4)]
+
+        assert all(t == q for t, q in zip(transform, expected))
+
+    # TODO: added with param shift
```
Comment (rmoyard) [inline tests/returntypes/test_spsa_gradient_new.py:406]: We should be able to add this one.

File: pennylane/gradients/spsa_gradient.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:271]: They are selected via `indices`, which is computed from the output of `choose_grad_methods` further up :) 

A bit of context (maybe this is clear already):
The way the `sampler` is supposed to work is that it obtains 
1. the indices within the list of trainable tape parameters that is supposed to be treated (e.g. trainable tape parameters that do not influence the output value of the QNode will be skipped)
2. the number of trainable tape parameters, including those that are not selected by `indices` in 1.
3. An iteration variable, which until now is mostly used in the dummy sampler in the test suite. `_rademacher_sampler` for example does not use this input at all.

and that it returns a `direction` vector that has the full size for all trainable tape parameters, but only the ones chosen by `indices` are actually filled with non-zero values.

In the tagged line itself we compute `_shifts`, which actually does _not_ have the full size of all trainable tape parameters, but just the same size as `indices`. This makes sure that the output Jacobian actually has the correct shape instead of returning zeros for the parameters that are not supposed to be differentiated.

File: pennylane/gradients/spsa_gradient.py (hunk 5 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:272]: Just to make sure: Do you mean why we use `inv_direction` over `direction`?
This can be seen as part of the definition of the SPSA gradient, see e.g. the equation labelled with **SP** in the beginning of [this SPSA Wikipedia entry](https://en.wikipedia.org/wiki/Simultaneous_perturbation_stochastic_approximation). Alternatively, you could consider SPSA to perform finite differences in all directions simultaneously, but with wrong evaluation points for each individual direction. In the gradient recipes for finite differences, we have a structure like
$\frac{1}{h} \sum_{i=1}^n a_i * f(h*x_i)$ for some coefficients $a_i$, evaluation shifts $x_i$ and the shift scale $h$. This division by $h$ in front basically is multiplication by `inv_direction`, just for the one-dimensional case.
I hope this makes sense :)

File: pennylane/gradients/spsa_gradient.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:287]: No, all results here should have the same shape, because they  only differ by the shifted _input_ values, and we are in the single-measurement case. But I might be missing some edge case here? Happy to discuss this further! :)

File: tests/gradients/test_spsa_gradient.py (hunk 11 above)
Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py:55]: Good point, thanks!

File: pennylane/gradients/spsa_gradient.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:107]: Gave it a shot, let me know whether it is clear like this :)

File: tests/gradients/test_spsa_gradient.py (hunk 11 above)
Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py:55]: I moved it and added a convergence test to the Rademacher test class.

File: tests/returntypes/test_spsa_gradient_new.py (hunk 12 above)
Comment (dwierichs) [reply tests/returntypes/test_spsa_gradient_new.py:406]: Yes, it will work regarding the support of special observables with SPSA. However, this test uses the `QNode` kwarg `diff_method`, which is not registered yet. I uncommented this test in all three test files and marked it with `pytest.mark.skip`. In #3440 the registration is added, and we will be able to add the tests correspondingly :)

File: pennylane/gradients/spsa_gradient.py [hunk 13]
```diff
@@ -0,0 +1,573 @@
...
+    if all(g == "0" for g in diff_methods):
+        return _all_zero_grad_new(tape, shots)
+
+    gradient_tapes = []
+    extract_r0 = False
+
+    coeffs, shifts = finite_diff_coeffs(n=n, approx_order=approx_order, strategy=strategy)
+
+    if 0 in shifts:
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:268]: You're right, the assumption is that `0` is the first shift in `shifts`. This is incorporated in the processing logic of parameter-shift rules as well as in `finite_diff_coeffs` and lets us skip any sorting steps after creating the differentiation recipes, and avoids indexing errors. If there is a place where additional documentation about this would be helpful, I'm happy for any suggestions you might have :)

File: pennylane/gradients/spsa_gradient.py (hunk 6 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py:308]: No I think you are correct!

File: pennylane/gradients/spsa_gradient.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/gradients/spsa_gradient.py:268]: > and avoids indexing errors

you mean to say that `shifts` will sometimes be an empty list, so this ensures we don't panic in that case?

File: pennylane/gradients/spsa_gradient.py (hunk 3 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py]: Yes we should update that in future PRs 👍

File: pennylane/gradients/spsa_gradient.py (hunk 4 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py:292]: Great explanation thank you!

File: tests/returntypes/test_spsa_gradient_new.py (hunk 12 above)
Comment (rmoyard) [reply tests/returntypes/test_spsa_gradient_new.py]: Ok thank you!

File: pennylane/gradients/spsa_gradient.py (hunk 4 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py:292]: @timmysilv Useful details here.

File: pennylane/gradients/spsa_gradient.py (hunk 5 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py:293]: Yes I was talking about `inv_direction`. Thanks for the link and explanation. Just to make sure, are we talking about $J(u_n -c_n \Delta_n)$?

File: pennylane/gradients/spsa_gradient.py [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+            checking.
+        shots (None, int, list[int], list[ShotTuple]): The device shots that will be used to
+            execute the tapes outputted by this transform. Note that this argument doesn't
+            influence the shots used for tape execution, but provides information
+            to the transform about the device shots and helps in determining if a shot
+            sequence was used to define the device shots for the new return types output system.
+        num_directions (int): Number of sampled simultaneous perturbation vectors. An estimate for
+            the gradient is computed for each vector using the underlying finite-difference
+            method, and afterwards all estimates are averaged.
+        sampler (callable): Sampling method to obtain the simultaneous perturbation directions.
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:117]: Clear and complete!

File: pennylane/gradients/spsa_gradient.py (hunk 5 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py:293]: (updated my comment)

File: pennylane/gradients/spsa_gradient.py (hunk 8 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py:164]: I think a `seed` keyword argument makes sense here, then we can pass it to the sampler. Do you see other parameters that would require kwargs for the sampler? If not I would go with the first option.

File: pennylane/gradients/spsa_gradient.py [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+                    qml.math.tensordot(qml.math.convert_like(_coeffs, res), res, axes=[[0], [0]])
+                    + grads
+                )
+            grads = grads / num_directions
+            if num_trainable_params == 1:
+                return qml.math.convert_like(grads[0], res)
+            return tuple(qml.math.convert_like(g, res) for g in grads)
+
+        grads = []
+        for i, _ in enumerate(tape.measurements):
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:319]:
```suggestion
        for i in range(len(tape.measurements)):
```
you can also create `num_measurement = len(tape.measurements)` as it is used multiple times.

File: pennylane/gradients/spsa_gradient.py [hunk 16]
```diff
@@ -0,0 +1,605 @@
...
+            grad = 0
+            for rep, _coeffs in enumerate(all_coeffs):
+                res = [r[i] for r in results[rep * tapes_per_grad : (rep + 1) * tapes_per_grad]]
+                if r0 is not None:
+                    res.insert(0, r0[i])
+                res = qml.math.stack(res)
```
Comment (rmoyard) [inline pennylane/gradients/spsa_gradient.py:325]: Double check: Here we might have multiple measurements with different shapes?

File: tests/returntypes/test_spsa_gradient_shot_vec_new.py [hunk 17]
```diff
@@ -0,0 +1,1394 @@
...
+        parameters"""
+        dev = qml.device("default.qubit", wires=2, shots=default_shot_vector)
+
+        weights = [0.1, 0.2]
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=0)
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        # TODO: remove once #2155 is resolved
```
Comment (rmoyard) [inline tests/returntypes/test_spsa_gradient_shot_vec_new.py:132]: Probably not relevant any more? (I hope)

File: tests/returntypes/test_spsa_gradient_shot_vec_new.py [hunk 18]
```diff
@@ -0,0 +1,1394 @@
...
+                    -(np.sin(x / 2) ** 2 * np.sin(y)) / 2,
+                ],
+                atol=spsa_shot_vec_tol,
+                rtol=0,
+            )
+
+
+@pytest.mark.parametrize("approx_order", [2])
+@pytest.mark.parametrize("strategy", ["center"])
+@pytest.mark.xfail(reason="TODO: higher-order derivatives with finite shots")
```
Comment (rmoyard) [inline tests/returntypes/test_spsa_gradient_shot_vec_new.py:967]: What is the limitation here? I think it works with finite-diff and shots.

File: tests/returntypes/test_spsa_gradient_new.py [hunk 19]
```diff
@@ -0,0 +1,1052 @@
...
+                qml.RY(x[1], wires=[1])
+                qml.CNOT(wires=[0, 1])
+                qml.expval(qml.PauliZ(0))
+                qml.probs(wires=[1])
+
+            tape.trainable_params = {0, 1}
+            tapes, fn = spsa_grad(tape, n=1, num_directions=num_directions, sampler=sampler)
+            jac = fn(dev.batch_execute(tapes))
+            if sampler == coordinate_sampler:
+                jac = tuple(tuple(2 * _j for _j in _jac) for _jac in jac)
```
Comment (rmoyard) [inline tests/returntypes/test_spsa_gradient_new.py:898]: What happens with the coordinate sampler for it to miss a factor 2?.

**Review by rmoyard — COMMENTED** [review]
Second round, it looks great already 💯  I have left some comments. There is an unexpected indentation and the doc build is failing: `None:27: WARNING: Unexpected indentation.` but it does help much.

File: pennylane/gradients/spsa_gradient.py (hunk 5 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:293]: Yes exactly, the $\Delta_n$ is the direction vector (of the $n$-th optimization iteration, in this case).

File: pennylane/gradients/spsa_gradient.py (hunk 8 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:164]: Okay, will do. Currently I do not see a concrete necessity. In the worst case, I guess the passed `sampler` would have to come with its `kwargs` fixed :)

File: pennylane/gradients/spsa_gradient.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:268]: Ah, no. I meant that if `0` was in `shifts`, but not first place, there would be quite a few functions in the `gradients` module that would have to find the index of the `0`, extract the corresponding coefficient correctly and so on... And that's a bit more error-prone I think.
Typically `shifts` is not empty, but if `0` is not in it (which happens rather often, actually), we don't have to extract an unshifted coefficient from the `coeffs`.

File: pennylane/gradients/spsa_gradient.py (hunk 16 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:325]: Yes, overall they might differ. But `res` is redefined for each `i`, i.e. for each return type shape, as far as I can tell.
So it should again be fine, just that the different `res` for different `i` might have different shapes (those are not stacked between the `i`s).

File: tests/returntypes/test_spsa_gradient_shot_vec_new.py (hunk 17 above)
Comment (dwierichs) [reply tests/returntypes/test_spsa_gradient_shot_vec_new.py:132]: #2155 is not resolved... :sweat_smile:

File: tests/returntypes/test_spsa_gradient_shot_vec_new.py (hunk 18 above)
Comment (dwierichs) [reply tests/returntypes/test_spsa_gradient_shot_vec_new.py:967]: It is copied from `test_finite_diff_shot_vec_new`. I suggest to resolve this in a separate PR.

File: tests/returntypes/test_spsa_gradient_new.py (hunk 19 above)
Comment (dwierichs) [reply tests/returntypes/test_spsa_gradient_new.py:898]: The concept of `SPSA` is to sample many directions, perform 1D finite differences for those, and average them to get an estimate for the gradient. If we want to make `SPSA` to actually compute the exact same gradient as `finite_diff`, i.e. sample each parameter direction successively, we instead want to compute the finite diff with the first direction being `[1, 0....0]` , the second `[0, 1, 0...0]` and so on, and sum them up.
The `mean` sums them up, but also divides by the number of directions, which we need to revert here, in order to just get the sum.

File: pennylane/gradients/spsa_gradient.py (hunk 8 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:164]: Done. I coded it up so that `sampler_seed` is passed to `sampler` for each of the `num_directions` iterations. I think for testing this makes sense, but it will basically destroy the stochastic averaging idea for `num_directions>1`, do we want that?
Alternatively, one could also just use `sampler_seed` in `_rademacher_sampler` if its the first direction, i.e. `idx_rep=0`. Let me know what you think! :)

File: pennylane/gradients/spsa_gradient.py [hunk 20]
```diff
@@ -0,0 +1,629 @@
...
+        >>> qml.gradients.param_shift(circuit)(params)
+        tensor([[-0.3875172 , -0.18884787, -0.38355704],
+                [ 0.69916862,  0.34072424,  0.69202359]], requires_grad=True)
+
+        As we can see, the SPSA output is a rather coarse approximation to the true
+        gradient, and this although the parameter-shift rule for three parameters uses
+        just six circuit evaluations, much fewer than SPSA! Consequentially, SPSA is
+        not necessarily useful for small circuits with few parameters, but will pay off
+        for large circuits where other gradient estimators require unfeasibly many circuit
+        executions.
```
Comment (timmysilv) [inline pennylane/gradients/spsa_gradient.py:501]: I'm not quite following what you're saying here. I thiiiink it's just the result of a typo but overall it sounds like you're saying parameter-shift is cheaper and more accurate, which kinda leaves me (aka the reader) wondering what the point SPSA is. I think the "although" is what is throwing me off 😄

**Review by timmysilv — APPROVED** [review]
this is awesome, thanks! I'd wait for Romain's approval, but it all looks good for me. My remaining questions are only about documentation.

File: pennylane/gradients/spsa_gradient.py [hunk 21]
```diff
@@ -0,0 +1,629 @@
...
+              the sampler.
+
+        sampler_seed (int or None): Seed passed to ``sampler``. The seed is passed in each
+            call to the sampler, so that only one unique direction is sampled even if
+            ``num_directions>1``.
+
+    Returns:
+        tensor_like or tuple[list[QuantumTape], function]:
+
+        - If the input is a QNode, a tensor
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:449]: 100% agree :)

File: pennylane/gradients/spsa_gradient.py (hunk 20 above)
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:501]: I rephrased this part, let me know what you think :)

File: pennylane/gradients/spsa_gradient.py (hunk 16 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py:338]: Thanks!

File: tests/returntypes/test_spsa_gradient_shot_vec_new.py (hunk 18 above)
Comment (rmoyard) [reply tests/returntypes/test_spsa_gradient_shot_vec_new.py:968]: Yes great

File: tests/returntypes/test_spsa_gradient_new.py (hunk 19 above)
Comment (rmoyard) [reply tests/returntypes/test_spsa_gradient_new.py:898]: Good to know, is it possible to add a comment about it somewhere?

File: pennylane/gradients/spsa_gradient.py (hunk 20 above)
Comment (rmoyard) [reply pennylane/gradients/spsa_gradient.py]: Thank you!

**Review by rmoyard — APPROVED** [review]
Awesome job @dwierichs 🥇 I let you close and resolve the last comments, otherwise it is good on my side.

File: pennylane/gradients/spsa_gradient.py (hunk 20 above)
Comment (timmysilv) [reply pennylane/gradients/spsa_gradient.py]: looks great, ty!

File: tests/returntypes/test_spsa_gradient_new.py (hunk 19 above)
Comment (dwierichs) [reply tests/returntypes/test_spsa_gradient_new.py:898]: Yes, I'll add this technical detail to the docstring of `spsa_grad` itself!

File: pennylane/gradients/spsa_gradient.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:449]: How do we keep track of this? Someone opens an issue?

[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:452]: Opened #3476 to fix this.
